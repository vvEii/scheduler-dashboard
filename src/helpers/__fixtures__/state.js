export function addInterview(id, interview) {
  return state => ({
    ...state,
    days: state.days.map(day =>
      day.appointments.includes(id) ? { ...day, spots: day.spots - 1 } : day
    ),
    appointments: {
      ...state.appointments,
      [id]: {
        ...state.appointments[id],
        interview: interview
      }
    }
  });
}

export default {
  days: [
    {
      id: 1,
      name: "Monday",
      appointments: [1, 2],
      interviewers: [1, 2],
      spots: 2
    },
    {
      id: 2,
      name: "Tuesday",
      appointments: [3, 4],
      interviewers: [1, 2],
      spots: 2
    }
  ],
  appointments: {
    "1": { id: 1, time: "12pm", interview: null },
    "2": { id: 2, time: "1pm", interview: null },
    "3": { id: 3, time: "12pm", interview: null },
    "4": { id: 4, time: "1pm", interview: null }
  },
  interviewers: {
    "1": {
      id: 1,
      name: "Sylvia Palmer",
      avatar: "https://i.imgur.com/LpaY82x.png"
    },
    "2": {
      id: 2,
      name: "Tori Malcolm",
      avatar: "https://i.imgur.com/Nmx0Qxo.png"
    }
  }
};
