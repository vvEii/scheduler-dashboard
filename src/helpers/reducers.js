import { filter } from "lodash/fp";

function getSpotsForDay(state, day) {
  const foundDay = state.days.find(d => d.name === day);

  if (!foundDay) throw new Error(`No Day Found: ${day}`);

  return filter(id => state.appointments[id].interview === null)(
    foundDay.appointments
  ).length;
}

export function setInterview(state, id, interview) {
  const appointments = {
    ...state.appointments,
    [id]: {
      ...state.appointments[id],
      interview
    }
  };

  return {
    ...state,
    days: state.days.map(day => ({
      ...day,
      spots: getSpotsForDay({ ...state, appointments }, day.name)
    })),
    appointments
  };
}
