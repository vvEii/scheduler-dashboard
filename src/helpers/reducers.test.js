import { flow } from "lodash/fp";

import { setInterview } from "./reducers";

import state, { addInterview } from "helpers/__fixtures__/state";

describe("selectors", () => {
  describe("setInterview", () => {
    it("should return state with new interview when interview is not null", () => {
      expect(
        setInterview(state, 2, {
          student: "Archie Cohen",
          interviewer: 1
        })
      ).toMatchSnapshot();
    });

    it("should return state with no interview when interview is not null", () => {
      const testState = JSON.parse(JSON.stringify(state));

      testState.appointments["2"].interview = {
        student: "Archie Cohen",
        interviewer: 1
      };

      expect(setInterview(testState, 2, null)).toMatchSnapshot();
    });

    it("should return state with one interview when interview is created and cancelled", () => {
      const testState = flow([
        addInterview(2, {
          student: "Archie Cohen",
          interviewer: 1
        }),
        addInterview(3, {
          student: "Lydia Miller-Jones",
          interviewer: 2
        })
      ])(state);

      expect(setInterview(testState, 3, null)).toMatchSnapshot();
    });
  });
});
