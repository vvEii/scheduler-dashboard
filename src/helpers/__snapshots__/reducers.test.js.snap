// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`selectors setInterview should return state with new interview when interview is not null 1`] = `
Object {
  "appointments": Object {
    "1": Object {
      "id": 1,
      "interview": null,
      "time": "12pm",
    },
    "2": Object {
      "id": 2,
      "interview": Object {
        "interviewer": 1,
        "student": "Archie Cohen",
      },
      "time": "1pm",
    },
    "3": Object {
      "id": 3,
      "interview": null,
      "time": "12pm",
    },
    "4": Object {
      "id": 4,
      "interview": null,
      "time": "1pm",
    },
  },
  "days": Array [
    Object {
      "appointments": Array [
        1,
        2,
      ],
      "id": 1,
      "interviewers": Array [
        1,
        2,
      ],
      "name": "Monday",
      "spots": 1,
    },
    Object {
      "appointments": Array [
        3,
        4,
      ],
      "id": 2,
      "interviewers": Array [
        1,
        2,
      ],
      "name": "Tuesday",
      "spots": 2,
    },
  ],
  "interviewers": Object {
    "1": Object {
      "avatar": "https://i.imgur.com/LpaY82x.png",
      "id": 1,
      "name": "Sylvia Palmer",
    },
    "2": Object {
      "avatar": "https://i.imgur.com/Nmx0Qxo.png",
      "id": 2,
      "name": "Tori Malcolm",
    },
  },
}
`;

exports[`selectors setInterview should return state with no interview when interview is not null 1`] = `
Object {
  "appointments": Object {
    "1": Object {
      "id": 1,
      "interview": null,
      "time": "12pm",
    },
    "2": Object {
      "id": 2,
      "interview": null,
      "time": "1pm",
    },
    "3": Object {
      "id": 3,
      "interview": null,
      "time": "12pm",
    },
    "4": Object {
      "id": 4,
      "interview": null,
      "time": "1pm",
    },
  },
  "days": Array [
    Object {
      "appointments": Array [
        1,
        2,
      ],
      "id": 1,
      "interviewers": Array [
        1,
        2,
      ],
      "name": "Monday",
      "spots": 2,
    },
    Object {
      "appointments": Array [
        3,
        4,
      ],
      "id": 2,
      "interviewers": Array [
        1,
        2,
      ],
      "name": "Tuesday",
      "spots": 2,
    },
  ],
  "interviewers": Object {
    "1": Object {
      "avatar": "https://i.imgur.com/LpaY82x.png",
      "id": 1,
      "name": "Sylvia Palmer",
    },
    "2": Object {
      "avatar": "https://i.imgur.com/Nmx0Qxo.png",
      "id": 2,
      "name": "Tori Malcolm",
    },
  },
}
`;

exports[`selectors setInterview should return state with one interview when interview is created and cancelled 1`] = `
Object {
  "appointments": Object {
    "1": Object {
      "id": 1,
      "interview": null,
      "time": "12pm",
    },
    "2": Object {
      "id": 2,
      "interview": Object {
        "interviewer": 1,
        "student": "Archie Cohen",
      },
      "time": "1pm",
    },
    "3": Object {
      "id": 3,
      "interview": null,
      "time": "12pm",
    },
    "4": Object {
      "id": 4,
      "interview": null,
      "time": "1pm",
    },
  },
  "days": Array [
    Object {
      "appointments": Array [
        1,
        2,
      ],
      "id": 1,
      "interviewers": Array [
        1,
        2,
      ],
      "name": "Monday",
      "spots": 1,
    },
    Object {
      "appointments": Array [
        3,
        4,
      ],
      "id": 2,
      "interviewers": Array [
        1,
        2,
      ],
      "name": "Tuesday",
      "spots": 2,
    },
  ],
  "interviewers": Object {
    "1": Object {
      "avatar": "https://i.imgur.com/LpaY82x.png",
      "id": 1,
      "name": "Sylvia Palmer",
    },
    "2": Object {
      "avatar": "https://i.imgur.com/Nmx0Qxo.png",
      "id": 2,
      "name": "Tori Malcolm",
    },
  },
}
`;
